### PR TITLE
Fix job scheduler

### LINKED ISSUE #809: JobScheduler running same job multiple times
When JobScheduler get reloaded by changing the file during runtime

Cron jobs will execute (or more).
(https://github.com/openmod/openmod/blob/12d263d6266c3e603b18d7ac1bee37bef60fcd00/framework/OpenMod.Core/Jobs/JobScheduler.cs#L321)
Soluction would be adding cancellation token.

Event jobs currently dont unsubscribe events, making it running twice(or more).

Delay jobs same as cron

### PR DESCRIPTION
**Pull request by rube200**
Fix #809

### CODE REVIEW

File: framework/OpenMod.Core/Helpers/AsyncHelper.cs [hunk 1]
```diff
@@ -68,7 +69,8 @@ public static void RunSync(Func<Task> action)
         /// <param name="name">The name of the task.</param>
         /// <param name="task">The task to run.</param>
         /// <param name="exceptionHandler">The optional exception handler.</param>
-        public static void Schedule(string name, Func<Task> task, Action<Exception>? exceptionHandler = null)
+        /// <param name="cancellationToken">The optional cancellation token.</param>
+        public static void Schedule(string name, Func<Task> task, Action<Exception>? exceptionHandler = null, CancellationToken? cancellationToken = null)
```
Comment (DiFFoZ) [inline framework/OpenMod.Core/Helpers/AsyncHelper.cs:73]: Breaking change!

**Review by DiFFoZ — CHANGES_REQUESTED** [review]

File: framework/OpenMod.Core/Jobs/JobScheduler.cs [hunk 2]
```diff
@@ -233,216 +225,208 @@ private async Task ScheduleJobInternalAsync(ScheduledJob job, bool isCalledFromS
...
-            m_ScheduledJobs.Add(job);
+            AddCancellableJob(job);
 
+            if (shouldRepeat)
+                m_Logger.LogInformation(
+                    "Scheduling job \"{JobName}\" type \"{JobType}\" with schedule \"{JobSchedule}\"", job.Name,
+                    job.Type, job.Schedule);
+            else
+                m_Logger.LogInformation("Delaying job \"{JobName}\" with delay of \"{JobDelay}\"", job.Name,
+                    $"{delay:c}");
```
Comment (Trojaner) [inline framework/OpenMod.Core/Jobs/JobScheduler.cs:301]: missing brackets

File: framework/OpenMod.Core/Jobs/JobScheduler.cs [hunk 3]
```diff
@@ -233,216 +225,208 @@ private async Task ScheduleJobInternalAsync(ScheduledJob job, bool isCalledFromS
...
+                    await RemoveJobAsync(job);
                 return;
             }
 
-            var delay = nextOccurence.Value - DateTimeOffset.Now;
-            if (delay.TotalMilliseconds <= 0)
-            {
+            var delay = GetJobDelay(job);
+            if (delay == null || delay.Value.TotalMilliseconds is < 0 or > int.MaxValue)
                 return;
```
Comment (Trojaner) [inline framework/OpenMod.Core/Jobs/JobScheduler.cs:291]: missing brackets

File: framework/OpenMod.Core/Jobs/JobScheduler.cs [hunk 4]
```diff
@@ -233,216 +225,208 @@ private async Task ScheduleJobInternalAsync(ScheduledJob job, bool isCalledFromS
...
+                            break;
 
-            if (eventNameDelimiterIndex == -1)
-            {
-                m_Logger.LogError("Invalid event job format \"{JobSchedule}\" for \"{JobName}\" job", job.Schedule, job.Name);
-                return;
-            }
+                        delay = GetJobDelay(job);
+                        if (delay!.Value.TotalMilliseconds is < 0 or > int.MaxValue)
+                            break;
```
Comment (Trojaner) [inline framework/OpenMod.Core/Jobs/JobScheduler.cs:325]: missing brackets

File: framework/OpenMod.Core/Jobs/JobScheduler.cs [hunk 5]
```diff
@@ -233,216 +225,208 @@ private async Task ScheduleJobInternalAsync(ScheduledJob job, bool isCalledFromS
...
+                {
+                    // ignore it
+                }
 
-            m_EventBus.Subscribe(m_Runtime, eventName, async (_, _, @event) => {
-                await ExecuteJobAsync(job, new { Event = @event });
+                if (removeAfterExec)
+                    await RemoveJobAsync(job);
+                else
+                    DisposeCancellableJob(job);
```
Comment (Trojaner) [inline framework/OpenMod.Core/Jobs/JobScheduler.cs:337]: missing brackets

File: framework/OpenMod.Core/Jobs/JobScheduler.cs [hunk 6]
```diff
@@ -233,216 +225,208 @@ private async Task ScheduleJobInternalAsync(ScheduledJob job, bool isCalledFromS
...
+            Exception? cronEx;
 
-            if (isNotDelayable)
+            try
             {
-                await ExecuteJobAsync(job);
+                var expression = CronExpression.Parse(job.Schedule);
+                var nextOccurence = expression.GetNextOccurrence(DateTimeOffset.Now, TimeZoneInfo.Local);
+                if (nextOccurence == null)
+                    return null;
```
Comment (Trojaner) [inline framework/OpenMod.Core/Jobs/JobScheduler.cs:350]: missing brackets

File: framework/OpenMod.Core/Jobs/JobScheduler.cs [hunk 7]
```diff
@@ -233,216 +225,208 @@ private async Task ScheduleJobInternalAsync(ScheduledJob job, bool isCalledFromS
...
+        private void SubscribeEventJob(ScheduledJob job)
         {
-            m_ScheduledJobs.Clear();
-            m_JobExecutors.Clear();
+            m_Logger.LogInformation("Subscribed job \"{JobName}\" to \"{JobSchedule}\" event", job.Name, job.Schedule);
+            var dispose = m_EventBus.Subscribe(m_OpenModComponent, job.Schedule!, async (_, _, @event) =>
+            {
+                var token = job.CancellationTokenSource!.Token;
+                if ((!job.Enabled ?? false) || !m_OpenModComponent.IsComponentAlive || token.IsCancellationRequested)
+                    return;
```
Comment (Trojaner) [inline framework/OpenMod.Core/Jobs/JobScheduler.cs:419]: missing brackets

File: framework/OpenMod.Core/autoexec.yaml [hunk 8]
```diff
@@ -33,7 +33,8 @@
 jobs:
   # Upgrade OpenMod every Sunday at 0:00 and then restart.
   - name: OpenMod Auto Update
-    schedule: '0 0 * * 0'
+    type: '@Repeat'
```
Comment (Trojaner) [inline framework/OpenMod.Core/autoexec.yaml:36]: lowercase

File: framework/OpenMod.Core/autoexec.yaml [hunk 9]
```diff
@@ -44,7 +45,8 @@ jobs:
 
   # Restart everyday at 3 AM.
   - name: Auto Restart
-    schedule: '0 3 * * *'
+    type: '@Repeat'
```
Comment (Trojaner) [inline framework/OpenMod.Core/autoexec.yaml:48]: lowercase

File: framework/OpenMod.Core/autoexec.yaml [hunk 10]
```diff
@@ -53,6 +55,7 @@ jobs:
 
   # Save every 5 minutes.
   - name: Auto save
+    type: '@Repeat'
```
Comment (Trojaner) [inline framework/OpenMod.Core/autoexec.yaml:58]: lowercase

File: framework/OpenMod.Core/autoexec.yaml [hunk 11]
```diff
@@ -73,7 +76,8 @@ jobs:
   # It works like terminal (Linux) or cmd (Windows),
   # so you can use scripts like sh or bat or directly run programs
   - name: Wipe
-    schedule: '0 0 1 * *'
+    type: '@Repeat'
```
Comment (Trojaner) [inline framework/OpenMod.Core/autoexec.yaml:79]: lowercase

File: framework/OpenMod.Core/Jobs/JobExecutorOptions.cs [hunk 12]
```diff
@@ -25,14 +21,9 @@ public JobExecutorOptions()
...
             if (!typeof(ITaskExecutor).IsAssignableFrom(type))
-            {
                 throw new Exception($"Type {type} must be an instance of IJobExecutor!");
-            }
 
-            if (m_JobExecutorTypes.Contains(type))
-            {
-                return;
-            }
+            if (m_JobExecutorTypes.Contains(type)) return;
```
Comment (Trojaner) [inline framework/OpenMod.Core/Jobs/JobExecutorOptions.cs:26]: missing brackets

File: framework/OpenMod.API/Jobs/ScheduledJob.cs [hunk 13]
```diff
@@ -1,57 +1,123 @@
...
+            if (ReferenceEquals(objA: null, obj)) return false;
+#pragma warning restore IDE0041 // Use '== null' or 'is null'
+
             if (ReferenceEquals(this, obj)) return true;
-            if (obj.GetType() != this.GetType()) return false;
-            return Equals((ScheduledJob)obj);
-        }
+            // ReSharper disable once ConvertIfStatementToReturnStatement
+            if (obj.GetType() != GetType())
+                return false;
```
Comment (Trojaner) [inline framework/OpenMod.API/Jobs/ScheduledJob.cs:118]: missing brackets

File: framework/OpenMod.API/Jobs/JobTask.cs [hunk 14]
```diff
@@ -4,25 +4,22 @@
...
         public JobTask(string jobName, string task, Dictionary<string, object?> args, params object[] parameters)
         {
             if (string.IsNullOrEmpty(jobName))
-            {
                 throw new ArgumentException(nameof(jobName));
-            }
 
             if (string.IsNullOrEmpty(task))
-            {
                 throw new ArgumentException(nameof(task));
```
Comment (Trojaner) [inline framework/OpenMod.API/Jobs/JobTask.cs:22]: missing brackets

**Review by Trojaner — CHANGES_REQUESTED** [review]

**Comment by Trojaner** [comment]
Why add `type: ` instead of using previous syntax? It's also a breaking change.

**Pull request merged**